README.md
# HydraDart
Hydra for [Dartlang](https://dart.dev) - A light-weight library for building distributed applications such as microservices.

![](HydraDart.png)

## Prior work
Hydra for Dart is based on the Hydra approach to building microservices.

See related repos at: https://github.com/pnxtech

This implementation for Dart is especially based on HydraPy for Python.

## Example

In this example HydraDart uses [Dart Shelf](https://pub.dev/packages/shelf) and [Dart Self Router](https://pub.dev/packages/shelf_router) to allow a Dart service to be discoverable within a Docker Swarm or Kubernetes cluster.

```dart
import 'dart:io';
import 'dart:convert';
import 'package:shelf_router/shelf_router.dart';
import 'package:shelf/shelf.dart';
import 'package:shelf/shelf_io.dart' as io;
import './hydra.dart';

class API {
  helloHandler(Request request) {
    return Response.ok('hello-world');
  }

  userHandler(Request request, String user) {
    return Response.ok('hello $user');
  }
}

void main(List<String> args) async {
  var hydra = Hydra();
  var router = Router();
  var api = API();

  hydra.bindRouter(router);
  hydra.addRoute('/v1/dart', 'get', api.helloHandler);
  hydra.addRoute('/v1/dart/user/<user>', 'get', api.userHandler);

  // Load configuration file
  File configFile = File('./configs/dart-svcs-config.json');
  Future<String> futureContent = configFile.readAsString();
  futureContent.then((config) async {
    Map<String, dynamic> configMap = jsonDecode(config);

    // Use any available host or container IP (usually `0.0.0.0`).
    final ip = InternetAddress.anyIPv4;

    // Configure a pipeline that logs requests.
    final routerHandler =
        Pipeline().addMiddleware(logRequests()).addHandler(router);

    // For running in containers, we respect the PORT environment variable.
    final port = configMap['hydra']['servicePort'];

    final server = await io.serve(routerHandler, ip, port);
    print('Server listening on port ${server.port}');

    hydra.init(configMap);
  });
}
```

In the example above our server application loads a `dart-svcs-config.json` file and passes it along to the `hydra.init()` member function.

```json
{
  "hydra": {
    "serviceName": "dart-svcs",
    "serviceIP": "",
    "servicePort": 7134,
    "serviceType": "test",
    "serviceDescription": "Dart experimental service",
    "plugins": {
      "hydraLogger": {
        "logToConsole": true,
        "onlyLogLocally": false
      }
    },
    "redis": {
      "urlxxx": "redis://redis:6379/15",
      "host": "redis",
      "port": 6379,
      "db": 15
    }
  }
}
```

CHANGELOG.md
# Change log

## 20220524 - Version 0.0.4
Async/Await improvments and a bug fix in the addRoute call.  The get switch handler had a return statement instead of a break.

## 20220517 - Version 0.0.1
This is the inital commit of this project.  Keep in mind that this project is experimental and likely to change over time.



lib/hydra_dart.dart
import 'dart:io';
import 'dart:async';
import 'dart:convert';
import 'package:redis/redis.dart';
import 'package:uuid/uuid.dart';
import 'package:shelf_router/shelf_router.dart';

/// HydraPresence is used for JSON conversion
///
class HydraPresence {
  String serviceName;
  String serviceDescription;
  String version;
  String instanceID;
  int processID;
  String ip;
  int port;
  String hostName;
  String updatedOn;
  HydraPresence(
      this.serviceName,
      this.serviceDescription,
      this.version,
      this.instanceID,
      this.processID,
      this.ip,
      this.port,
      this.hostName,
      this.updatedOn);
  Map toJson() => {
        'serviceName': serviceName,
        'serviceDescription': serviceDescription,
        'version': version,
        'instanceID': instanceID,
        'processID': processID,
        'ip': ip,
        'port': port,
        'hostName': hostName,
        'updatedOn': updatedOn
      };
}

/// HydraMemory is used with HydraHealth to define memory values
///
class HydraMemory {
  int rss;
  int heapTotal;
  int heapUsed;
  int ext;
  int arrayBuffers;
  HydraMemory(
      this.rss, this.heapTotal, this.heapUsed, this.ext, this.arrayBuffers);
  Map toJson() => {
        'rss': rss,
        'heapTotal': heapTotal,
        'heapUsed': heapUsed,
        'external': ext,
        'arrayBuffers': arrayBuffers
      };
}

/// HydraHealth is used to generate the JSON for serialization to
/// Redis
///
class HydraHealth {
  String updatedOn;
  String serviceName;
  String instanceID;
  String hostName;
  String sampledOn;
  int processID;
  String architecture;
  String platform;
  String nodeVersion;
  HydraMemory hydraMemory;
  HydraHealth(
      this.updatedOn,
      this.serviceName,
      this.instanceID,
      this.hostName,
      this.sampledOn,
      this.processID,
      this.architecture,
      this.platform,
      this.nodeVersion,
      this.hydraMemory);
  Map toJson() => {
        'updatedOn': updatedOn,
        'serviceName': serviceName,
        'instanceID': instanceID,
        'hostName': hostName,
        'sampledOn': sampledOn,
        'processID': processID,
        'architecture': architecture,
        'platform': platform,
        'nodeVersion': nodeVersion,
        'memory': hydraMemory,
        'uptimeSeconds': 0 // current time - uptime
      };
}

/// UMF
///
class UMF {
  String _to = '';
  String _frm = '';
  String _mid = '';
  String _ts = '';
  String _bdy = '';

  UMF(String to, String frm, String bdy) {
    _to = to;
    _frm = frm;
    _mid = (Uuid()).v4();
    _ts = timeStamp();
    _bdy = bdy;
  }
  toJsonString() {
    String json = '''{
      "to": "$_to",
      "frm": "$_frm",
      "mid": "$_mid",
      "ts": "$_ts",
      "bdy": $_bdy
    }''';
    return jsonEncode(jsonDecode(json));
  }
}

/// timeStamp
///
timeStamp() {
  String s = DateTime.now().toUtc().toIso8601String();
  return '${s.substring(0, s.length - 4)}Z';
}

/// Hydra is the main module used to support microservices
/// functionality using Redis
///
class Hydra {
  static const int oneSecond = 1;
  static const int oneWeekInSeconds = 604800;
  static const int presenceUpdateInterval = oneSecond;
  static const int heatlhUpdateInterval = oneSecond * 5;
  static const int keyExperationTTL = oneSecond * 3;
  static const String redisPreKey = 'hydra:service';
  static const String mcMessageKey = 'hydra:service:mc';
  late RedisConnection redis;
  late int redisDB;
  late Command redisCommand;
  late Timer periodicTimer;
  late String serverInstanceID;
  int healthTick = 0;
  late String serviceName;
  late String serviceDescription;
  late String hostName;
  late int processID;
  late String ip;
  late int port;
  late String version = '0.0.1';

  late Router routerInstance;
  List<String> hydraRoutes = [];

  /// init is used to initialize the Hydra module
  /// init accepts a map which is created from reading a JSON based
  /// configuration file
  ///
  init(Map<String, dynamic> configMap) async {
    ip = await getLocalIP();
    port = configMap['hydra']['servicePort'];
    hostName = Platform.localHostname;
    processID = pid;
    serviceName = configMap['hydra']['serviceName'];
    serviceDescription = configMap['hydra']['serviceDescription'];
    var uuid = Uuid();
    serverInstanceID = uuid.v4().replaceAll(RegExp('-'), '');

    periodicTimer =
        Timer.periodic(const Duration(seconds: oneSecond), heartBeat);

    redisDB = configMap['hydra']['redis']['db'];
    redis = RedisConnection();
    redisCommand = await redis.connect(configMap['hydra']['redis']['host'],
        configMap['hydra']['redis']['port']);
    await registerRoutes();
  }

  /// shutdown
  ///
  shutdown() {
    periodicTimer.cancel();
    redis.close();
  }

  /// heartBeat
  ///
  heartBeat(timer) async {
    healthTick++;
    if (healthTick == heatlhUpdateInterval) {
      healthTick = 0;
      var response = await redisCommand.send_object(['SELECT', redisDB]);
      if (response == 'OK') {
        await redisCommand.send_object([
          'SETEX',
          '$redisPreKey:$serviceName:$serverInstanceID:health',
          keyExperationTTL,
          health()
        ]);
      }
    }

    var response = await redisCommand.send_object(['SELECT', redisDB]);
    if (response == 'OK') {
      await redisCommand.send_object([
        'SETEX',
        '$redisPreKey:$serviceName:$serverInstanceID:presence',
        keyExperationTTL,
        serverInstanceID
      ]);
      await redisCommand.send_object(
          ['HSET', '$redisPreKey:nodes', serverInstanceID, presence()]);
    }
  }

  /// presence
  ///
  presence() {
    return jsonEncode(HydraPresence(serviceName, serviceDescription, version,
        serverInstanceID, processID, ip, port, hostName, timeStamp()));
  }

  /// health
  ///
  health() {
    HydraMemory memory = HydraMemory(0, 0, 0, 0, 0);
    String ts = timeStamp();
    return jsonEncode(HydraHealth(ts, serviceName, serverInstanceID, hostName,
        ts, pid, 'dart', 'linux', version, memory));
  }

  /// getLocalIP
  ///
  Future getLocalIP() async {
    List<String> ips = [];
    for (var interface in await NetworkInterface.list()) {
      for (var addr in interface.addresses) {
        if (addr.type.name == 'IPv4' && !addr.isLoopback) {
          if (!addr.address.startsWith('172.')) {
            ips.insert(0, addr.address);
          }
        }
      }
    }
    return ips[0];
  }

  /// flushRoutes
  ///
  flushRoutes() async {
    var response = await redisCommand.send_object(['SELECT', redisDB]);
    if (response == 'OK') {
      await redisCommand.send_object(
          ['DEL', '$redisPreKey:$serviceName:$serverInstanceID:routes']);
    }
  }

  /// bindRouter
  ///
  bindRouter(Router router) {
    routerInstance = router;
  }

  /// addRoute
  ///
  addRoute(String path, String method, Function handler) {
    switch (method) {
      case 'get':
        routerInstance.get(path, handler);
        break;
      case 'post':
        routerInstance.post(path, handler);
        break;
    }
    var transformedPath = path.replaceAll('<', ':').replaceAll('>', '');
    hydraRoutes.insert(0, '[$method]$transformedPath');
  }

  /// registerRoutes
  ///
  registerRoutes() async {
    await flushRoutes();
    addRoute('/$serviceName', 'get', () => {});
    addRoute('/$serviceName/', 'get', () => {});
    addRoute('/$serviceName/:rest', 'get', () => {});

    UMF umf = UMF('hydra-router:/refresh', '$serviceName:/', '''
      {
        "action": "refresh",
        "serviceName": "$serviceName"
      }
    ''');

    var response = await redisCommand.send_object(['SELECT', redisDB]);
    if (response == 'OK') {
      for (final route in hydraRoutes) {
        await redisCommand.send_object(
            ['SADD', '$redisPreKey:$serviceName:service:routes', route]);
      }
      await redisCommand.send_object(
          ['PUBLISH', '$mcMessageKey:hydra-router', umf.toJsonString()]);
    }
  }
}


lib/hydra.dart
import 'dart:io';
import 'dart:async';
import 'dart:convert';
import 'dart:mirrors';
import 'package:redis/redis.dart';
import 'package:uuid/uuid.dart';
import 'package:shelf_router/shelf_router.dart';

/// HydraPresence is used for JSON conversion
///
class HydraPresence {
  String serviceName;
  String serviceDescription;
  String version;
  String instanceID;
  int processID;
  String ip;
  int port;
  String hostName;
  String updatedOn;
  HydraPresence(
      this.serviceName,
      this.serviceDescription,
      this.version,
      this.instanceID,
      this.processID,
      this.ip,
      this.port,
      this.hostName,
      this.updatedOn);
  Map toJson() => {
        'serviceName': serviceName,
        'serviceDescription': serviceDescription,
        'version': version,
        'instanceID': instanceID,
        'processID': processID,
        'ip': ip,
        'port': port,
        'hostName': hostName,
        'updatedOn': updatedOn
      };
}

/// HydraMemory is used with HydraHealth to define memory values
///
class HydraMemory {
  int rss;
  int heapTotal;
  int heapUsed;
  int ext;
  int arrayBuffers;
  HydraMemory(
      this.rss, this.heapTotal, this.heapUsed, this.ext, this.arrayBuffers);
  Map toJson() => {
        'rss': rss,
        'heapTotal': heapTotal,
        'heapUsed': heapUsed,
        'external': ext,
        'arrayBuffers': arrayBuffers
      };
}

/// HydraHealth is used to generate the JSON for serialization to
/// Redis
///
class HydraHealth {
  String updatedOn;
  String serviceName;
  String instanceID;
  String hostName;
  String sampledOn;
  int processID;
  String architecture;
  String platform;
  String nodeVersion;
  HydraMemory hydraMemory;
  HydraHealth(
      this.updatedOn,
      this.serviceName,
      this.instanceID,
      this.hostName,
      this.sampledOn,
      this.processID,
      this.architecture,
      this.platform,
      this.nodeVersion,
      this.hydraMemory);
  Map toJson() => {
        'updatedOn': updatedOn,
        'serviceName': serviceName,
        'instanceID': instanceID,
        'hostName': hostName,
        'sampledOn': sampledOn,
        'processID': processID,
        'architecture': architecture,
        'platform': platform,
        'nodeVersion': nodeVersion,
        'memory': hydraMemory,
        'uptimeSeconds': 0 // current time - uptime
      };
}

/// UMF
///
class UMF {
  String _to = '';
  String _frm = '';
  String _mid = '';
  String _ts = '';
  String _bdy = '';

  UMF(String to, String frm, String bdy) {
    _to = to;
    _frm = frm;
    _mid = (Uuid()).v4();
    _ts = timeStamp();
    _bdy = bdy;
  }
  toJsonString() {
    String json = '''{
      "to": "$_to",
      "frm": "$_frm",
      "mid": "$_mid",
      "ts": "$_ts",
      "bdy": $_bdy
    }''';
    return jsonEncode(jsonDecode(json));
  }
}

/// timeStamp
///
timeStamp() {
  String s = DateTime.now().toUtc().toIso8601String();
  return '${s.substring(0, s.length - 4)}Z';
}

/// Hydra is the main module used to support microservices
/// functionality using Redis
///
class Hydra {
  static const int oneSecond = 1;
  static const int oneWeekInSeconds = 604800;
  static const int presenceUpdateInterval = oneSecond;
  static const int heatlhUpdateInterval = oneSecond * 5;
  static const int keyExperationTTL = oneSecond * 3;
  static const String redisPreKey = 'hydra:service';
  static const String mcMessageKey = 'hydra:service:mc';
  late RedisConnection redis;
  late int redisDB;
  late Command redisCommand;
  late Timer periodicTimer;
  late String serverInstanceID;
  int healthTick = 0;
  late String serviceName;
  late String serviceDescription;
  late String hostName;
  late int processID;
  late String ip;
  late int port;
  late String version = '0.0.1';

  late Router routerInstance;
  List<String> hydraRoutes = [];

  /// init is used to initialize the Hydra module
  /// init accepts a map which is created from reading a JSON based
  /// configuration file
  ///
  init(Map<String, dynamic> configMap) async {
    ip = await getLocalIP();
    port = configMap['hydra']['servicePort'];
    hostName = Platform.localHostname;
    processID = pid;
    serviceName = configMap['hydra']['serviceName'];
    serviceDescription = configMap['hydra']['serviceDescription'];
    var uuid = Uuid();
    serverInstanceID = uuid.v4().replaceAll(RegExp('-'), '');

    periodicTimer =
        Timer.periodic(const Duration(seconds: oneSecond), heartBeat);

    redisDB = configMap['hydra']['redis']['db'];
    redis = RedisConnection();
    redisCommand = await redis.connect(configMap['hydra']['redis']['host'],
        configMap['hydra']['redis']['port']);
    await registerRoutes();
  }

  /// shutdown
  ///
  shutdown() {
    periodicTimer.cancel();
    redis.close();
  }

  /// heartBeat
  ///
  heartBeat(timer) {
    healthTick++;
    if (healthTick == heatlhUpdateInterval) {
      healthTick = 0;
      redisCommand.send_object(['SELECT', redisDB]).then((var response) {
        if (response == 'OK') {
          redisCommand.send_object([
            'SETEX',
            '$redisPreKey:$serviceName:$serverInstanceID:health',
            keyExperationTTL,
            health()
          ]);
        }
      });
    }

    redisCommand.send_object(['SELECT', redisDB]).then((var response) {
      if (response == 'OK') {
        redisCommand.send_object([
          'SETEX',
          '$redisPreKey:$serviceName:$serverInstanceID:presence',
          keyExperationTTL,
          serverInstanceID
        ]);
        redisCommand.send_object(
            ['HSET', '$redisPreKey:nodes', serverInstanceID, presence()]);
      }
    });
  }

  /// presence
  ///
  presence() {
    return jsonEncode(HydraPresence(serviceName, serviceDescription, version,
        serverInstanceID, processID, ip, port, hostName, timeStamp()));
  }

  /// health
  ///
  health() {
    HydraMemory memory = HydraMemory(0, 0, 0, 0, 0);
    String ts = timeStamp();
    return jsonEncode(HydraHealth(ts, serviceName, serverInstanceID, hostName,
        ts, pid, 'dart', 'linux', version, memory));
  }

  /// getLocalIP
  ///
  Future getLocalIP() async {
    List<String> ips = [];
    for (var interface in await NetworkInterface.list()) {
      for (var addr in interface.addresses) {
        if (addr.type.name == 'IPv4' && !addr.isLoopback) {
          if (!addr.address.startsWith('172.')) {
            ips.insert(0, addr.address);
          }
        }
      }
    }
    return ips[0];
  }

  /// flushRoutes
  ///
  flushRoutes() async {
    await redisCommand.send_object(['SELECT', redisDB]).then((var response) {
      if (response == 'OK') {
        redisCommand.send_object(
            ['DEL', '$redisPreKey:$serviceName:$serverInstanceID:routes']);
      }
    });
  }

  /// bindRouter
  ///
  bindRouter(Router router) {
    routerInstance = router;
  }

  /// addRoute
  ///
  addRoute(String path, String method, Function handler) {
    // https://stackoverflow.com/questions/13293345/dynamic-class-method-invocation-in-dart
    // var im = reflect(router);
    // im.invoke(method, [handler]);

    // TODO: Replace below with the use of dart annotations and reflection
    // https://www.youtube.com/watch?v=_2yjPLVEGs4
    switch (method) {
      case 'get':
        routerInstance.get(path, handler);
        return;
      case 'post':
        routerInstance.post(path, handler);
        break;
    }
    var transformedPath = path.replaceAll('<', ':').replaceAll('>', '');
    hydraRoutes.insert(0, '[$method]$transformedPath');
  }

  /// registerRoutes
  ///
  registerRoutes() async {
    await flushRoutes();
    addRoute('/$serviceName', 'get', () => {});
    addRoute('/$serviceName/', 'get', () => {});
    addRoute('/$serviceName/:rest', 'get', () => {});

    UMF umf = UMF('hydra-router:/refresh', '$serviceName:/', '''
      {
        "action": "refresh",
        "serviceName": "$serviceName"
      }
    ''');

    await redisCommand.send_object(['SELECT', redisDB]).then((var response) {
      if (response == 'OK') {
        for (final route in hydraRoutes) {
          redisCommand.send_object(
              ['SADD', '$redisPreKey:$serviceName:service:routes', route]);
        }
        redisCommand.send_object(
            ['PUBLISH', '$mcMessageKey:hydra-router', umf.toJsonString()]);
      }
    });
  }
}
